core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java

```diff
@@ -239,4 +261,62 @@ private IndexMetaData addDefaultUnitsIfNeeded(IndexMetaData indexMetaData) {
         // No changes:
         return indexMetaData;
     }
+
+
+    /**
+     * Checks the mappings of the mappings for compatibility with the current version
+     */
+    private IndexMetaData upgradeMappings(IndexMetaData indexMetaData) {
+        Index index = new Index(indexMetaData.getIndex());
+        Settings settings = indexMetaData.settings();
+        try {
+            SimilarityLookupService similarityLookupService = new SimilarityLookupService(index, settings);
+            // We cannot instantiate real analysis server at this point because the node might not have
+            // been started yet. However, we don't really need real analyzers at this stage - so we can fake it
+            try (AnalysisService analysisService = new FakeAnalysisService(index, settings)) {
+                try (MapperService mapperService = new MapperService(index, settings, analysisService, similarityLookupService, scriptService)) {
+                    for (ObjectCursor<MappingMetaData> cursor : indexMetaData.getMappings().values()) {
+                        MappingMetaData mappingMetaData = cursor.value;
+                        mapperService.merge(mappingMetaData.type(), mappingMetaData.source(), true, true);
```

Comment:
Actually I think applyDefaults should be false too since we already have a mapping?